### ISSUE TITLE
ActivatorUtilities fails if first constructor is unresolvable

### ISSUE THREAD

**Issue by pardahlman** [aspnet/DependencyInjection#464]
I want to report a bug for the static method `CreateInstance(IServiceProvider provider, Type instanceType, params object[] parameters)` on `ActivatorUtilities`.

When trying to resolve a service that has unresolvable argument(s) in its first declared constructor, an `InvalidOperationException` is thrown (_Unable to resolve service for type 'IDependency' while attempting to activate 'Service'._)

To reproduce, create the following classes

```csharp
public class Service
{
    public Service(IUnknownDependency unknown) { }
    public Service(IKnownDependency known) { }
}

public interface IUnknownDependency { }

public interface IKnownDependency { }

public class KnownDependency : IKnownDependency { }
```
Create a service collection and register the `IKnownDependency ` and `Service`, build the service provider and try to resolve `Service`

```csharp
var provider = new ServiceCollection()
	.AddTransient<IKnownDependency, KnownDependency>()
	.AddSingleton<Service>()
	.BuildServiceProvider();
var service = ActivatorUtilities.CreateInstance(provider, typeof(Service)); // throws!
```
However, if the order of the constructors are changed in `Service`, the call to `CreateInstance` returns the service.

**Comment by pranavkm** [comment]
This is essentially a dup of https://github.com/aspnet/DependencyInjection/issues/197. We might address this in 1.2.0 by adding an additional flag to the method that says use a slower best match algorithm.

**Comment by pardahlman** [comment]
You are right, it looks like the same problem 👍 . However, in #197, the problem discussed is how to identify the "best" constructor, in this scenario it is more about finding an applicable constructor/avoiding an exception to be thrown. But OK - just wanted to bring this to your attention.

**Comment by pranavkm** [comment]
The problem is the same though. The service container is opaque to `ActivatorUtilities`. Consequently we have no way of telling if `IUnknownDependency` exists in the container (short of calling `serviceProvider.GetService<IUnknownDependency>()`). Our design choice here was to run with the first ctor available. #197 proposes fixing this by possibly calling `GetService` until a best matching ctor is found. In your case, it would be the 2nd one.

That said, if you DIed in the type, rather than use a service location (which is what `ActivatorUtilities` does), you shouldn't see this. The default service provider (the one you get from `BuildServiceProvider`) supports finding the best matching ctor. Something to consider in case you can refactor your code.

**Comment by pardahlman** [comment]
Understood 👌 .

I'm writing a middleware inspired library where users can register their middleware

```csharp
pipe.Use<QueueDeclareMiddleware>(new QueueDeclareOptions())
```
Where `QueueDeclareOptions` is not registered in the DI, but uniquely provided for each use.  `QueueDeclareMiddleware` in turn has constructor parameters that are resolved from DI. The `IServiceProvider` interface does not have a method where I can provide additional arguments, which is why I use the `ActivatorUtilities` to start with. Do you have any recommendations for this particular scenario?

**Comment by pranavkm** [comment]
cc @Tratcher - wouldn't this work out of the box if your ctor accepts an `IOptions<QueueDeclareOptions>`? The middleware pipeline should do the constructor injection for you while injecting the option as `IOptions<T>`. Maybe I haven't followed your scenario clearly.

**Comment by pardahlman** [comment]
Perhaps - I haven't looked at the options framework since RC2. The way I understand the usage of `IOptions<T>` is to provide a way to apply multiple `Action<T>` to the same options object. Is this correct?

Let me explain my scenario a bit more, and tell me if you think the Options framework can help me. A typical "middleware chain" could look something like this

```csharp
pipe
	.Use<StageMarkerMiddleware>(StageMarkerOptions.For(PublishStage.Initiated))
	.Use<PublishConfigurationMiddleware>()
	.Use<StageMarkerMiddleware>(StageMarkerOptions.For(PublishStage.ExchangeConfigured))
	.Use<ExchangeDeclareMiddleware>()
	.Use<StageMarkerMiddleware>(StageMarkerOptions.For(PublishStage.ExchangeDeclared))
	.Use<RoutingKeyMiddleware>()
	.Use<StageMarkerMiddleware>(StageMarkerOptions.For(PublishStage.RoutingKeyCreated)); 
```
As you can see, the `StageMarkerMiddleware` is declared multiple times, each with its own `StageMarkerOption`. The arguments for the call to `Use<TMiddleware>(params objects[] args)`  are stored away in an object to allow me to instantiate the middleware when the pipe is actually built.
```csharp
public class MiddlewareInfo
{
	public Type Type { get; set; }
	public object[] ConstructorArgs { get; set; }
}
```
The middleware may have other constructor arguments that are provided by the DI. I was hoping to find someway to provide additional parameters to the `IServiceProvider` like this
```csharp
var service = serviceProvider.GetService(info.Type, info.ConstructorArgs);
```
But I didn't find any way to do so. Looking at the [test for the Options framework](https://github.com/aspnet/Options/blob/dev/test/Microsoft.Extensions.Options.Test/OptionsTest.cs#L143) i get the feeling that it applies the action to the same `TOptions` object, and when the service provider is built, it will return a single options object which has passed through all the "configurations" of it. Correct?

Looking at the [Options example from the documentation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration#options-example) it also looks like I need to wrap my Options class in a `IOptions<T>` wrapper. To me it feels a bit like it is violating the separation of concern, as this is something that is specific to the DI framework. Talking about DI frameworks, I try to be agnostic in the core part of the lib, so I'm a bit hesitant to include something that is service collection / service provider specific.

Thanks for your help!

**Comment by pranavkm** [comment]
The `IOptions` approach wouldn't work here - it passes around a singleton instance around, so you wouldn't get different instances of `StageMarkerOptions` at different stages with it. @Tratcher \ @davidfowl might have other suggestions for you.

**Comment by Tratcher** [comment]
It should still work for this specific case, the options instance passed into UseMiddleware should take precedence over resolving an options instance from DI.

**Comment by pardahlman** [comment]
@Tratcher Ok sounds good, but how would I register/use the specific argument passed to `UseMiddleware` to achieve that? It doesn't make sense for me to register it in the service collection, as the instances provided are "scoped" for that particular instance of the middleware. It would be very helpful with a few lines of code explaining

1. What I would need to with with the arguments passed into `UseMiddleware<T>(params object []args) to register (or not register, not sure I follow what you mean) the instance in the service collection
2. What I would need to do to be able to resolve the middleware with the provided arguments from the service provider.

(To solve this problem in Autofac, I would probably [pass the arguments to the resolver](http://docs.autofac.org/en/latest/resolve/parameters.html).)

**Comment by Tratcher** [comment]
1. Should just work, have you tried it yet?. The args you pass into UseMiddleware are used along side the service container to activate the middleware, they are not added to the container. https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs#L72-L75 
2. I don't think this is possible with multiple instances of the type, at least not with our container.

**Comment by pardahlman** [comment]
Maybe I was a bit unclear earlier. I'm not referring to the `Use<TMiddleware>` of `IApplicationBuilder`, but to [a pipe builder](https://github.com/pardahlman/RawRabbit/blob/2.0/src/RawRabbit/Pipe/PipeBuilder.cs#L14) that I've developed (heavily inspired by the application builder).

Looking at the `UseMiddlewareExtensions` you mentioned, it [also uses the `ActivatorUtilities`](https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs#L75), which was the original topic of this issue. @pranavkm then advised me against the `ActivatorUtilities` in order to find a suiting constructor.

> if you DIed in the type, rather than use a service location (which is what ActivatorUtilities does), you shouldn't see this.

I think that [this comment](https://github.com/aspnet/DependencyInjection/issues/464#issuecomment-260121791) describes my problem pretty well (I've updated it for clarity).

If I understand you correctly, the default service collection / service provider does not have support for resolving a service based on registered services in combination with additional provided constructor arguments. This can only be achieved by using `ActivatorUtilities`, that in turn tries to instantiate it with the first constructor, which in turn may throw (as per this ticket).
